telnet.c

```diff
@@ -29,7 +29,26 @@
 
 static ssize_t telnet_write(struct ios_ops *ios, const void *buf, size_t count)
 {
-	return write(ios->fd, buf, count);
+	size_t handled = 0;
+	ssize_t ret;
+	void *iac;
+
+	while ((iac = memchr(buf + handled, IAC, count - handled)) != NULL) {
+		if (iac - (buf + handled)) {
+			ret = write(ios->fd, buf + handled, iac - (buf + handled));
+			if (ret < 0)
+				return ret;
+			handled += ret;
+		} else {
+			dprintf(ios->fd, "%c%c", IAC, IAC);
```

Comment:
Will `dprintf` handle `EINTR` and incomplete writes correctly?

It took me some time to understand what this code should do. Perhaps it would be better to first quote the IAC in a local buffer (which can't fail) and then use a generic retrying send helper?

Reply:
Incomplete writes are a principle problem before and after my patch also in the other backends. See also issue #21.

The problem with quoting the IAC first is that the buffer is provided by `mux.c`, so the telnet code would need to allocate memory (which complicates the code further).

I added a few comments to hopefully simplify understanding the code.

Reply:
Why must only the first IAC be quoted?

Reply:
> Why must only the first IAC be quoted?
Hmm, should I add "Iterate for all IACs in the input" to the comment? In the code all `IAC`s are handled.

Reply:
Ah, the comment confused me. Being explicit about iteration sounds good.

I still think this would be simpler by just allocating count*2, handling the quoting and then one write. :)

Reply:
> I still think this would be simpler by just allocating count*2, handling the quoting and then one write. :)

I found another difficulty with this approach: I want to keep the return code semantic as is. This is necessary to later properly handle partial writes. If `telnet_write` is called with (say) 40 bytes to write, there are 3 `IAC`s in the input, so in the end `write(ios->fd, localbuf, 43);` is called. If that write returns 20, it's not trivial to determine what to return. In the end this has the downside to allocate extra memory and isn't easier than the approach I chose now.

Reply:
The idea is to not block in `.write()` to keep microcom responsive to user input (and maybe other events) and so I want to do completion of writes in the core instead of in the callback directly.
Once this works, `.write()` can also become simpler and look e.g. like this:

    if (count == 0)
        return 0;
    if (buf[0] == IAC) {
        return write_full(ios->fd, "%c%c", IAC, IAC); // TODO: better handle partial write
    } else {
        const char *iac = memchr(buf, IAC, count);
        if (!iac)
            iac = buf + count;
        return write(ios->fd, buf, iac - buf);
    }

Reply:
Ah. As long as you may need to write more than one byte in the lower level write, you may need to block. Perhaps split the quoting from the write function? The quoting function would append to a buffer and the core could handle draining the buffer asynchronously (i.e. via `poll()`/`select()`), perhaps using `libuv`/`libevent`?

Reply:
Yeah, the eventual plan is something like that. However I hesitate to add a dependency like libuv or libevent, so I tend to do it directly with `poll()` or `select()` (or maybe `epoll()`).